### PR TITLE
add feature - export particle objects as multimeshinstance to godot

### PR DESCRIPTION
**Pull request by epth**
It has some limitations for now, if a particle-system use a texture to scale particle size,the size in Godot multimesh will be incorrect

### CODE REVIEW

**Comment by epth** [comment]
This is The Test File, should enable the  Particle Enum-Button at the Top while exporting

[particle2multimesh_test.zip](https://github.com/godotengine/godot-blender-exporter/files/5001578/particle2multimesh_test.zip)

File: io_scene_godot/converters/multimesh.py
Comment (rcorre) [inline io_scene_godot/converters/multimesh.py:11]: comment should be updated.

File: io_scene_godot/converters/multimesh.py
Comment (rcorre) [inline io_scene_godot/converters/multimesh.py:22]: Names are already normalized here: https://github.com/godotengine/godot-blender-exporter/blob/564aa4f56cfe72be2d3c76893fe0380bb7548b8e/io_scene_godot/structures.py#L163-L164

If there's additional normalization needed it should probably be added there, unless it is specific to particles.

File: io_scene_godot/converters/multimesh.py
Comment (rcorre) [inline io_scene_godot/converters/multimesh.py:27]: remove these commented lines?

File: io_scene_godot/converters/multimesh.py
Comment (rcorre) [inline io_scene_godot/converters/multimesh.py:43]: prefer `format` to `%` for forward-compatibility.

File: io_scene_godot/converters/simple_nodes.py
Comment (rcorre) [inline io_scene_godot/converters/simple_nodes.py:20]: remove these prints before merging

File: io_scene_godot/converters/__init__.py
Comment (rcorre) [inline io_scene_godot/converters/__init__.py:43]: missing newline at end of file

File: io_scene_godot/converters/multimesh.py
Comment (rcorre) [inline io_scene_godot/converters/multimesh.py:82]: again, prefer `format` to `%`

**Review by rcorre — COMMENTED** [review]
I'm not in charge of this project, but I gave it a quick review since I'm interested in this feature!

I recently had two use cases: scattering lillypads across water and stalactites in a cave. Both were easier to do using particles in blender.

You've got a number of linter warnings, check these out: https://travis-ci.org/github/godotengine/godot-blender-exporter/jobs/714011504#L311

Great work though!

File: io_scene_godot/converters/simple_nodes.py
Comment (epth) [reply io_scene_godot/converters/simple_nodes.py:20]: thanks for reviewing!I have no much experience on Github, so which way is the best to remove there prints?

Comment (epth) [reply io_scene_godot/converters/simple_nodes.py]: 2 commits has pushed.

Comment (rcorre) [reply io_scene_godot/converters/simple_nodes.py]: Looks like you figured it out 👍

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [inline io_scene_godot/converters/multimesh.py:140]: Sorry, I think I need some context here 😅
Are these quaternion operations handles the up-axis difference between Godot and Blender?

**Review by Jason0214 — COMMENTED** [review]
Thank you very much! This is a great functionality to be added. I have played with it and it works great!  I am trying to understand how the particle settings being exported, I am still missing some pieces, so I left some comments. Hope you can help me a little bit and get it merged soon. Thanks again!

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:17]: good idea!

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:140]: no,its' not difference between Godot and Blender,but is difference between Blender's particle and Blender's object,i have tested many times,i thought this may because particles in Blender are very low level and computed directly by GPU,dont know too much about that, just guess....

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:173]: Hi,what if change the mat4_to_string in structure.py to 
```
def mat4_to_string(mtx, prefix='Transform(', suffix=')'):
    """Converts a matrix to a "Transform" string that can be parsed by Godot"""
    mtx = fix_matrix(mtx)
    array = Array(prefix,suffix)
```

and CONVERSIONS is still like this, is this correct?
```
CONVERSIONS = {
    float: float_to_string,
    bool: lambda x: 'true' if x else 'false',
    mathutils.Matrix: mat4_to_string,
```

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:153]: not sure it's correct,just tested out,may beacause of unknown param in Blender's particle system but not found the param

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [reply io_scene_godot/converters/multimesh.py:173]: Good idea! From my understanding of python, I think it works. Even if not, the CI would catch the issue.

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [reply io_scene_godot/converters/multimesh.py:140]: Okay. We can use the current implementation. We will figure it out when people start to use this feature.

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [reply io_scene_godot/converters/multimesh.py:153]: same as above.

File: io_scene_godot/converters/multimesh.py
Comment (Calinou) [reply io_scene_godot/converters/multimesh.py]: Is `%` actually deprecated and going to be removed in a future version? [pyformat.info](https://pyformat.info/) doesn't mention that.

Comment (rcorre) [reply io_scene_godot/converters/multimesh.py]: This was surprisingly hard to dig up: https://docs.python.org/2/library/stdtypes.html#str.format

> This method of string formatting is the new standard in Python 3, and should be preferred to the % formatting described in String Formatting Operations in new code.

And https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting: 

> The formatting operations described here exhibit a variety of quirks that lead to a number of common errors (such as failing to display tuples and dictionaries correctly). Using the newer formatted string literals, the str.format() interface, or template strings may help avoid these errors. Each of these alternatives provides their own trade-offs and benefits of simplicity, flexibility, and/or extensibility.

So it seems generally warned against, but probably nothing terrible will happen if you decide to use it 🤷

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:153]: yeah

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:18]: In particle system params, If "Render - Render As" are switched to "Collection",there maybe are several objects instanced to one particle,but in Godot one MultiMeshInstance just have one object to instance, so  choose the first object in Blender to display as the only one object in Godot's MultiMeshInstance's resource.

And the instance_object stand for "Render - Render As"  "Object" option,this is most similar with Godot's MultiMeshInstance. 

Don't know if Godot will support multiple object instance to one MultiMeshInstance LOL.

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:66]: same as above,Because if "Render-Render As" are switched to "Collection",there maybe several objects instanced in one particle system.so when we choose one of particle system,maybe different object should be instance,I admit just use hash from Bpy is good for now. I thought too much maybe give an option that can select which object in "collection" to export in future haha.

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [inline io_scene_godot/converters/multimesh.py:38]: I think you mean `i==ob.particle_systems.active` here?  In this loop, `i` is an instance of type `ParticleSystem`.

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [inline io_scene_godot/converters/multimesh.py:18]: maybe use `for ps in ob.particle_systems` here?  which is more concise.

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [reply io_scene_godot/converters/multimesh.py]: Thank you very much for the explanation. Now I have some idea of how it works. It does seem that a inadequacy in Godot to make a particle with multiple meshes. 

I agree it makes sense for now to use the first mesh here. 
Could you add a line of comment  here to explain the reason that we only export `collection[0]`. It would be very helpful. Thanks

File: io_scene_godot/converters/multimesh.py
Comment (Jason0214) [reply io_scene_godot/converters/multimesh.py:72]: Very good point, if in the future we have custom logic for each object referring an instanced particle system to export to a MutltiMeshInstance, then a bpy ParticleSystem hash might not be enough.

I am okay with current implementation and I think it works well. However, I still feel that the `MultiMeshResourceKey` is a little bit overcomplicated. In my opinion, make the key refer to bpy ParticleSystem object instead of the particle system name is better. The later is guaranteed unique. The `gd_rsc_type` seems unnecessary, only `MultiMesh` is involved here.

Anyway I will merge it whether you change it or not.

**Review by Jason0214 — APPROVED** [review]
Btw, could you also fix up the pycodestyle warning as well. Really appreciate it. : )

```
pycodestyle io_scene_godot
io_scene_godot/structures.py:425:25: E231 missing whitespace after ','
io_scene_godot/converters/multimesh.py:17:23: E225 missing whitespace around operator
io_scene_godot/converters/multimesh.py:26:80: E501 line too long (82 > 79 characters)
io_scene_godot/converters/multimesh.py:31:80: E501 line too long (80 > 79 characters)
io_scene_godot/converters/multimesh.py:37:1: W293 blank line contains whitespace
io_scene_godot/converters/multimesh.py:38:13: E225 missing whitespace around operator
io_scene_godot/converters/multimesh.py:39:31: E225 missing whitespace around operator
io_scene_godot/converters/multimesh.py:175:55: E231 missing whitespace after ','
io_scene_godot/converters/multimesh.py:175:65: E231 missing whitespace after ','
Makefile:12: recipe for target 'pep8' failed
``` 

You can check the Travis CI status, see https://travis-ci.org/github/godotengine/godot-blender-exporter/jobs/741590402

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py]: sure

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:38]: yeah,i will fix it soon

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:18]: you are right!

File: io_scene_godot/converters/multimesh.py
Comment (epth) [reply io_scene_godot/converters/multimesh.py:72]: Thanks and not change for now!

**Comment by Jason0214** [comment]
Thank you very much！It looks better now.
Though we still get some code style [warning from pycodestyle](https://travis-ci.org/github/godotengine/godot-blender-exporter/jobs/742198071). The rules are rigid 😆, but we have been using it until now, it is easier to fix the code a little bit to make the linter happy.
- You can make variable names a little bit longer and meaningful
- For the duplicated code, if you can not make it pass, I am okay with just disable it. Check the other examples in the code, I think you can find out how to do it. 
Btw, you can click the ❌ or ✔️ near your every commit to see the Travis CI build log.
Really appreciate your effort, I think we are very clock to merge 🎉

**Comment by epth** [comment]
@Jason0214 
have no idea with how to fix error of duplicated code.

**Comment by Jason0214** [comment]
My bad. Looks like disable pylint's `duplicate-code` disable is [broken](https://github.com/PyCQA/pylint/issues/214) : (
I pushed a commit to just using bpy ParticleSystem as the hash key, it is not the most robust one, but enough be working for now. Don't have enough time to wait for the CI build. I will check it later.
If you are okay with the change, I will merge this PR after CI is fixed.

**Comment by epth** [comment]
> 
> 
> My bad. Looks like disable pylint's `duplicate-code` disable is [broken](https://github.com/PyCQA/pylint/issues/214) : (
> I pushed a commit to just using bpy ParticleSystem as the hash key, it is not the most robust one, but enough be working for now. Don't have enough time to wait for the CI build. I will check it later.
> If you are okay with the change, I will merge this PR after CI is fixed.

Im ok, you can merge it !

**Pull request merged**
